include/llvm-dialects/Dialect/Dialect.td

```diff
@@ -197,15 +197,40 @@ def AttrI64 : IntegerAttr<"uint64_t", I64>;
 
 class Trait;
 
-class LlvmAttributeTrait<string llvmEnum_> : Trait {
+class LlvmEnumAttributeTrait<string llvmEnum_> : Trait {
   string llvmEnum = llvmEnum_;
 }
 
-def NoUnwind : LlvmAttributeTrait<"NoUnwind">;
-def ReadNone : LlvmAttributeTrait<"ReadNone">;
-def ReadOnly : LlvmAttributeTrait<"ReadOnly">;
-def WillReturn : LlvmAttributeTrait<"WillReturn">;
+def NoUnwind : LlvmEnumAttributeTrait<"NoUnwind">;
+def WillReturn : LlvmEnumAttributeTrait<"WillReturn">;
 
+/// Represent the LLVM `memory(...)` attribute as the OR (or union) of memory
+/// effects. An empty effects list means the operation does not access memory
+/// at all. If this trait is missing, it means that there is no bound on the
+/// memory accesses performed by the operation.
+///
+/// Each memory effect is given as a DAG `($modref location, ...)`. If the list
+/// of memory locations is empty, it means all memory may be accessed in this
+/// way.
+///
+/// Example: `Memory<[(ref), (mod ArgMem, InaccessibleMem)]>` means the
+/// operation may read from any kind of memory and write to argument and
+/// inaccessible memory.
+class Memory<list<dag> effects_> : Trait {
+  list<dag> effects = effects_;
+}
+
+def ref;
+def mod;
+def modref;
```

Comment:
Maybe read, write and readwrite is easier to understand? (I now found these defined in [MemoryEffects](https://llvm.org/doxygen/namespacellvm.html#a9db8c44b250b90e3ab7e4d144e7c9c2e); as you said offline, the textual IR calls them read and write)